File: src/test/java/no/nav/fo/veilarbregistrering/oppfolging/adapter/OppfolgingClientTest.java
package no.nav.fo.veilarbregistrering.oppfolging.adapter;

import com.google.common.net.MediaType;
import no.nav.common.featuretoggle.UnleashService;
import no.nav.common.sts.SystemUserTokenProvider;
import no.nav.fo.veilarbregistrering.FileToJson;
import no.nav.fo.veilarbregistrering.autorisasjon.AutorisasjonService;
import no.nav.fo.veilarbregistrering.bruker.AktorId;
import no.nav.fo.veilarbregistrering.bruker.Bruker;
import no.nav.fo.veilarbregistrering.bruker.Foedselsnummer;
import no.nav.fo.veilarbregistrering.metrics.MetricsService;
import no.nav.fo.veilarbregistrering.profilering.Innsatsgruppe;
import no.nav.fo.veilarbregistrering.profilering.Profilering;
import no.nav.fo.veilarbregistrering.profilering.ProfileringService;
import no.nav.fo.veilarbregistrering.registrering.bruker.*;
import no.nav.fo.veilarbregistrering.sykemelding.SykemeldingService;
import no.nav.fo.veilarbregistrering.sykemelding.adapter.SykemeldingGatewayImpl;
import no.nav.fo.veilarbregistrering.sykemelding.adapter.SykmeldtInfoClient;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockserver.integration.ClientAndServer;

import javax.inject.Provider;
import javax.servlet.http.HttpServletRequest;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.mockserver.model.HttpRequest.request;
import static org.mockserver.model.HttpResponse.response;

class OppfolgingClientTest {
    private static final String MOCKSERVER_URL = "localhost";
    private static final int MOCKSERVER_PORT = 1084;

    private static final Foedselsnummer IDENT = Foedselsnummer.of("10108000398"); //Aremark fiktivt fnr.";
    private static final Fnr FNR = new Fnr("10108000398"); //Aremark fiktivt fnr.";
    private static final Bruker BRUKER = Bruker.of(IDENT, AktorId.of("AKTØRID"));

    private InaktivBrukerService inaktivBrukerService;
    private OppfolgingClient oppfolgingClient;
    private ClientAndServer mockServer;

    @AfterEach
    public void tearDown() {
        mockServer.stop();
    }

    @BeforeEach
    public void setup() {

        mockServer = ClientAndServer.startClientAndServer(MOCKSERVER_PORT);
        oppfolgingClient = buildClient();
        BrukerRegistreringRepository brukerRegistreringRepository = mock(BrukerRegistreringRepository.class);
        SykmeldtInfoClient sykeforloepMetadataClient = mock(SykmeldtInfoClient.class);
        ProfileringService profileringService = mock(ProfileringService.class);
        UnleashService unleashService = mock(UnleashService.class);
        AutorisasjonService autorisasjonService = mock(AutorisasjonService.class);
        MetricsService metricsService = mock(MetricsService.class);

        OppfolgingGatewayImpl oppfolgingGateway = new OppfolgingGatewayImpl(oppfolgingClient);

        BrukerTilstandService brukerTilstandService = new BrukerTilstandService(
                oppfolgingGateway,
                new SykemeldingService(new SykemeldingGatewayImpl(sykeforloepMetadataClient), autorisasjonService, metricsService), unleashService);

        inaktivBrukerService = new InaktivBrukerService(
                brukerTilstandService,
                brukerRegistreringRepository,
                oppfolgingGateway);

        when(profileringService.profilerBruker(anyInt(), any(), any()))
                .thenReturn(new Profilering()
                        .setInnsatsgruppe(Innsatsgruppe.STANDARD_INNSATS)
                        .setAlder(50)
                        .setJobbetSammenhengendeSeksAvTolvSisteManeder(true));
    }

    private OppfolgingClient buildClient() {
        SystemUserTokenProvider systemUserTokenProvider = mock(SystemUserTokenProvider.class);
        Provider<HttpServletRequest> httpServletRequestProvider = mock(Provider.class);
        HttpServletRequest httpServletRequest = mock(HttpServletRequest.class);
        when(httpServletRequestProvider.get()).thenReturn(httpServletRequest);
        when(httpServletRequest.getHeader(any())).thenReturn("");
        when(systemUserTokenProvider.getSystemUserToken()).thenReturn("testToken");
        String baseUrl = "http://" + MOCKSERVER_URL + ":" + MOCKSERVER_PORT;
        UnleashService unleashService = mock(UnleashService.class);
        when(unleashService.isEnabled(any())).thenReturn(false);
        return this.oppfolgingClient = new OppfolgingClient(baseUrl, systemUserTokenProvider);
    }

    @Test
    public void skal_kaste_RuntimeException_ved_vilkaarlig_feil_som_ikke_er_haandtert() {
        mockServer.when(request().withMethod("POST").withPath("/oppfolging/aktiverbruker")).respond(response().withStatusCode(404));
        assertThrows(RuntimeException.class, () -> oppfolgingClient.aktiverBruker(new AktiverBrukerData(FNR, Innsatsgruppe.SITUASJONSBESTEMT_INNSATS)));
    }

    @Test
    public void skal_returnere_response_ved_204() {
        mockServer.when(request().withMethod("POST").withPath("/oppfolging/aktiverbruker")).respond(response().withStatusCode(204).withBody(okRegistreringBody(), MediaType.JSON_UTF_8));

        assertNotNull(oppfolgingClient.aktiverBruker(new AktiverBrukerData(FNR, Innsatsgruppe.SITUASJONSBESTEMT_INNSATS)));
    }

    @Test
    public void testAtReaktiveringFeilerDersomArenaSierAtBrukerErUnderOppfolging() {
        mockServer.when(request().withMethod("GET").withPath("/oppfolging"))
                .respond(response().withBody(settOppfolgingOgReaktivering(true, false), MediaType.JSON_UTF_8).withStatusCode(200));

        assertThrows(RuntimeException.class, () -> inaktivBrukerService.reaktiverBruker(BRUKER));
    }

    @Test
    public void testAtReaktiveringGirOKDersomArenaSierAtBrukerKanReaktiveres() {
        mockServer.when(request().withMethod("GET").withPath("/oppfolging"))
                .respond(response().withBody(settOppfolgingOgReaktivering(false, true), MediaType.JSON_UTF_8).withStatusCode(200));
        mockServer.when(request().withMethod("POST").withPath("/oppfolging/reaktiverbruker")).respond(response().withStatusCode(204));

        inaktivBrukerService.reaktiverBruker(BRUKER);
    }

    //FIXME: Vurder å flytte disse testene til en ny testklasse med mindre kompleksitet hvor vi tester Client eller
    // Gateway, men ikke alt annet samtidig.
    /*
    @Test
    public void testAtGirInternalServerErrorExceptionDersomBrukerIkkkeHarTilgangTilOppfolgingStatus() {
        mockServer.when(request().withMethod("GET").withPath("/oppfolging")).respond(response().withStatusCode(401));
        assertThrows(InternalServerErrorException.class, () -> brukerRegistreringService.hentStartRegistreringStatus(BRUKER, null));
    }

    @Test
    public void testAtGirInternalServerErrorExceptionDersomOppfolgingFeiler() {
        mockServer.when(request().withMethod("GET").withPath("/oppfolging")).respond(response().withStatusCode(500));
        assertThrows(InternalServerErrorException.class, () -> brukerRegistreringService.hentStartRegistreringStatus(BRUKER, null));
    }

    @Test
    public void testAtGirWebApplicationExceptionExceptionDersomIngenTilgang() {
        mockServer.when(request().withMethod("GET").withPath("/oppfolging")).respond(response().withStatusCode(403));
        assertThrows(WebApplicationException.class, () -> brukerRegistreringService.hentStartRegistreringStatus(BRUKER, null));
    }
*/
    @Test
    public void skaL_returnere_ved_manglende_oppholdstillatelse() {
        mockServer.when(request().withMethod("POST").withPath("/oppfolging/aktiverbruker")).respond(response().withBody(FileToJson.toJson("/oppfolging/manglerOppholdstillatelse.json")).withStatusCode(403));

        AktiverBrukerResultat actual = oppfolgingClient.aktiverBruker(new AktiverBrukerData(FNR, Innsatsgruppe.SITUASJONSBESTEMT_INNSATS));

        assertThat(actual.feil()).isEqualTo(AktiverBrukerFeil.BRUKER_MANGLER_ARBEIDSTILLATELSE);
    }

    //FIXME: Vurder å flytte disse testene til en ny testklasse med mindre kompleksitet hvor vi tester Client eller
    // Gateway, men ikke alt annet samtidig.
    /*
    @Test
    public void testAtGirIngenExceptionsDersomKun200OK() {
        mockServer.when(request().withMethod("GET").withPath("/oppfolging"))
                .respond(response().withBody(settOppfolgingOgReaktivering(true, false), MediaType.JSON_UTF_8).withStatusCode(200));

        assertNotNull(brukerRegistreringService.hentStartRegistreringStatus(BRUKER));
    }

    @Test
    public void testAtGirIngenExceptionsDersomKun200MedKanReaktiveresNull() {
        mockServer.when(request().withMethod("GET").withPath("/oppfolging"))
                .respond(response().withBody(settOppfolgingOgReaktivering(null, null), MediaType.JSON_UTF_8).withStatusCode(200));

        when(arbeidsforholdGateway.hentArbeidsforhold(any())).thenReturn(flereArbeidsforholdTilfeldigSortert());

        assertNotNull(brukerRegistreringService.hentStartRegistreringStatus(BRUKER));
    }
*/

    private String settOppfolgingOgReaktivering(Boolean oppfolging, Boolean reaktivering) {
        return "{\"kanReaktiveres\": "+reaktivering+", \"underOppfolging\": "+oppfolging+"}";
    }

    private String okRegistreringBody() {
        return "{\n" +
                "\"Status\": \"STATUS_SUKSESS\"\n" +
                "}";
    }
}

File: src/test/java/no/nav/fo/veilarbregistrering/oppfolging/adapter/OppfolgingClientTest.kt
package no.nav.fo.veilarbregistrering.oppfolging.adapter

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkStatic
import no.nav.common.sts.SystemUserTokenProvider
import no.nav.fo.veilarbregistrering.FileToJson
import no.nav.fo.veilarbregistrering.bruker.Foedselsnummer
import no.nav.fo.veilarbregistrering.config.RequestContext
import no.nav.fo.veilarbregistrering.profilering.Innsatsgruppe
import no.nav.fo.veilarbregistrering.registrering.bruker.AktiverBrukerException
import no.nav.fo.veilarbregistrering.registrering.bruker.AktiverBrukerFeil
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.mockserver.integration.ClientAndServer
import org.mockserver.junit.jupiter.MockServerExtension
import org.mockserver.model.HttpRequest
import org.mockserver.model.HttpResponse
import org.mockserver.model.MediaType
import javax.servlet.http.HttpServletRequest
import javax.ws.rs.core.HttpHeaders

@ExtendWith(MockServerExtension::class)
internal class OppfolgingClientTest(private val mockServer: ClientAndServer) {
    private lateinit var oppfolgingClient: OppfolgingClient


    @BeforeEach
    fun setup() {
        mockServer.reset()
        oppfolgingClient = buildClient(jacksonObjectMapper().findAndRegisterModules())
    }

    private fun buildClient(objectMapper: ObjectMapper): OppfolgingClient {
        mockkStatic(RequestContext::class)
        val systemUserTokenProvider: SystemUserTokenProvider = mockk()
        val httpServletRequest: HttpServletRequest = mockk()
        every { RequestContext.servletRequest() } returns httpServletRequest
        every { httpServletRequest.getHeader(HttpHeaders.COOKIE) } returns "czas Å\u009Brodkowoeuropejski standardowy"
        every { systemUserTokenProvider.systemUserToken } returns "testToken"
        val baseUrl = "http://" + mockServer.remoteAddress().address.hostName + ":" + mockServer.remoteAddress().port
        return OppfolgingClient(objectMapper, mockk(relaxed = true), baseUrl, systemUserTokenProvider) { "TOKEN" }.also { oppfolgingClient = it }
    }

    @Test
    fun skal_kaste_RuntimeException_ved_vilkaarlig_feil_som_ikke_er_haandtert() {
        mockServer.`when`(HttpRequest.request().withMethod("POST").withPath("/oppfolging/aktiverbruker")).respond(
            HttpResponse.response().withStatusCode(404)
        )
        assertThrows<RuntimeException> {
            oppfolgingClient.aktiverBruker(
                AktiverBrukerData(
                    FNR,
                    Innsatsgruppe.SITUASJONSBESTEMT_INNSATS
                )
            )
        }
    }

    @Test
    fun skal_returnere_response_ved_204() {
        mockServer.`when`(HttpRequest.request().withMethod("POST").withPath("/oppfolging/aktiverbruker")).respond(
            HttpResponse.response()
                    .withStatusCode(204)
                    .withBody(okRegistreringBody(), MediaType.JSON_UTF_8)
        )
        Assertions.assertNotNull(
            oppfolgingClient.aktiverBruker(
                AktiverBrukerData(
                    FNR,
                    Innsatsgruppe.SITUASJONSBESTEMT_INNSATS
                )
            )
        )
    }

    @Test
    fun `skal returnere respons for hentOppfolgingstatus`() {
        mockServer.`when`(HttpRequest.request().withMethod("GET").withPath("/oppfolging")).respond(
            HttpResponse.response()
                    .withStatusCode(200)
                    .withBody(ikkeUnderOppfolgingBody(), MediaType.JSON_UTF_8)
        )
        Assertions.assertNotNull(
            oppfolgingClient.hentOppfolgingsstatus(Foedselsnummer.of(FNR.fnr))
        )
    }

    @Test
    fun testAtReaktiveringFeilerDersomArenaSierAtBrukerErUnderOppfolging() {
        mockServer.`when`(HttpRequest.request().withMethod("GET").withPath("/oppfolging"))
            .respond(
                HttpResponse.response().withBody(settOppfolgingOgReaktivering(true, false), MediaType.JSON_UTF_8)
                    .withStatusCode(200)
            )
        assertThrows<RuntimeException> {
            oppfolgingClient.reaktiverBruker(Fnr("10108000398")) }
    }

    @Test
    fun testAtReaktiveringGirOKDersomArenaSierAtBrukerKanReaktiveres() {
        mockServer.`when`(HttpRequest.request().withMethod("GET").withPath("/oppfolging"))
            .respond(
                HttpResponse.response().withBody(settOppfolgingOgReaktivering(false, true), MediaType.JSON_UTF_8)
                    .withStatusCode(200)
            )
        mockServer.`when`(HttpRequest.request().withMethod("POST").withPath("/oppfolging/reaktiverbruker")).respond(
            HttpResponse.response().withStatusCode(204)
        )
        oppfolgingClient.reaktiverBruker(Fnr("10108000398"))
    }

    @Test
    fun `skal kaste riktig feil ved manglende oppholdstillatelse`() {
        mockServer.`when`(HttpRequest.request().withMethod("POST").withPath("/oppfolging/aktiverbruker")).respond(
            HttpResponse.response().withBody(FileToJson.toJson("/oppfolging/manglerOppholdstillatelse.json"))
                .withStatusCode(403)
        )
        val exception: AktiverBrukerException = assertThrows {
            oppfolgingClient.aktiverBruker(AktiverBrukerData(FNR, Innsatsgruppe.SITUASJONSBESTEMT_INNSATS))
        }
        assertThat(exception.aktiverBrukerFeil).isEqualTo(AktiverBrukerFeil.BRUKER_MANGLER_ARBEIDSTILLATELSE)
    }

    @Test
    fun `skal kaste riktig feil dersom bruker ikke kan reaktiveres`() {
        mockServer.`when`(HttpRequest.request().withMethod("POST").withPath("/oppfolging/aktiverbruker")).respond(
            HttpResponse.response().withBody(FileToJson.toJson("/oppfolging/kanIkkeReaktiveres.json"))
                .withStatusCode(403)
        )
        val exception: AktiverBrukerException = assertThrows {
            oppfolgingClient.aktiverBruker(AktiverBrukerData(FNR, Innsatsgruppe.SITUASJONSBESTEMT_INNSATS))
        }
        assertThat(exception.aktiverBrukerFeil).isEqualTo(AktiverBrukerFeil.BRUKER_KAN_IKKE_REAKTIVERES)
    }

    private fun settOppfolgingOgReaktivering(oppfolging: Boolean, reaktivering: Boolean): String {
        return "{\"kanReaktiveres\": $reaktivering, \"underOppfolging\": $oppfolging}"
    }

    private fun ikkeUnderOppfolgingBody(): String = """
        { "underOppfolging": false }
    """.trimIndent()

    private fun okRegistreringBody(): String {
        return """
            {
            "Status": "STATUS_SUKSESS"
            }
            """.trimIndent()
    }

    companion object {
        private val FNR = Fnr("10108000398") //Aremark fiktivt fnr.";
    }
}